### PR TITLE
[RISCV] Add stack clash protection

### PR DESCRIPTION
**Pull request by rzinsly**
Enable `-fstack-clash-protection` for RISCV and stack probe for function prologues.
We probe the stack by creating a loop that allocates and probe the stack in ProbeSize chunks.
We emit an unrolled probe loop for small allocations and emit a variable length probe loop for bigger ones.

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/stack-clash-prologue-nounwind.ll [hunk 1]
```diff
@@ -0,0 +1,208 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -mtriple=riscv64 -mattr=+m,+v -O2 < %s \
+; RUN:   | FileCheck %s -check-prefix=RV64I
```
Comment (lenary) [inline llvm/test/CodeGen/RISCV/stack-clash-prologue-nounwind.ll:3]: Why is this only testing rv64?

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 2]
```diff
@@ -610,22 +610,134 @@ static MCCFIInstruction createDefCFAOffset(const TargetRegisterInfo &TRI,
                                         Comment.str());
 }
 
+// Allocate stack space and probe it if necessary.
 void RISCVFrameLowering::allocateStack(MachineBasicBlock &MBB,
                                        MachineBasicBlock::iterator MBBI,
-                                       StackOffset Offset, bool EmitCFI,
-                                       unsigned CFIIndex) const {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:616]: I see the CFIIndex argument was removed. Can that be done as a separate patch?

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 3]
```diff
@@ -1939,3 +2049,72 @@ bool RISCVFrameLowering::isSupportedStackID(TargetStackID::Value ID) const {
...
 }
+
+// Synthesize the probe loop.
+static void emitStackProbeInline(MachineFunction &MF, MachineBasicBlock &MBB,
+                                 MachineBasicBlock::iterator MBBI,
+                                 DebugLoc DL) {
+
+  auto &Subtarget = MF.getSubtarget<RISCVSubtarget>();
+  const RISCVInstrInfo *TII = Subtarget.getInstrInfo();
+  bool IsRV64 = Subtarget.hasFeature(RISCV::Feature64Bit);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:2060]: Subtarget.is64Bit()?

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 4]
```diff
@@ -610,22 +610,134 @@ static MCCFIInstruction createDefCFAOffset(const TargetRegisterInfo &TRI,
...
+          .setMIFlag(MachineInstr::FrameSetup);
+    }
+
+    return;
+  }
+
+  // Unroll the probe loop depending on the number of iterations.
+  if (Offset < ProbeSize * 5) {
+    uint64_t CurrentOffset = 0;
+    bool IsRV64 = STI.hasFeature(RISCV::Feature64Bit);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:644]: STI.is64Bit()?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 5]
```diff
@@ -22156,3 +22156,25 @@ namespace llvm::RISCVVIntrinsicsTable {
...
+  // If the function specifically requests inline stack probes, emit them.
+  if (MF.getFunction().hasFnAttribute("probe-stack"))
+    return MF.getFunction().getFnAttribute("probe-stack").getValueAsString() ==
+           "inline-asm";
+
+  return false;
+}
+
+unsigned RISCVTargetLowering::getStackProbeSize(const MachineFunction &MF,
+                                                unsigned StackAlign) const {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:22171]: Use `Align StackAlign`.

File: llvm/lib/Target/RISCV/RISCVMachineFunctionInfo.cpp [hunk 6]
```diff
@@ -26,6 +27,35 @@ MachineFunctionInfo *RISCVMachineFunctionInfo::clone(
...
+    ProbeSize = F.getFnAttributeAsParsedInteger("stack-probe-size");
+  else if (const auto *PS = mdconst::extract_or_null<ConstantInt>(
+               F.getParent()->getModuleFlag("stack-probe-size")))
+    ProbeSize = PS->getZExtValue();
+  assert(int64_t(ProbeSize) > 0 && "Invalid stack probe size");
+
+  // Round down to the stack alignment.
+  uint64_t StackAlign =
+      STI->getFrameLowering()->getTransientStackAlign().value();
+  ProbeSize = std::max(StackAlign, ProbeSize & ~(StackAlign - 1U));
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVMachineFunctionInfo.cpp:47]: Can we use `Align::commonAlignment` instead of `ProbeSize & ~(StackAlign - 1U)`?

Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVMachineFunctionInfo.cpp:47]: Can we use `std::max(StackAlign, alignDown(ProbeSize, StackAlign))`?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 7]
```diff
@@ -22156,3 +22156,25 @@ namespace llvm::RISCVVIntrinsicsTable {
...
+  // The default stack probe size is 4096 if the function has no
+  // stack-probe-size attribute.
+  const Function &Fn = MF.getFunction();
+  unsigned StackProbeSize =
+      Fn.getFnAttributeAsParsedInteger("stack-probe-size", 4096);
+  // Round down to the stack alignment.
+  StackProbeSize &= ~(StackAlign - 1);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:22178]: Use `alignDown` from MathExtras.h?

**Comment by rzinsly** [comment]
@topperc thanks for the review, I followed your suggestions.

**Comment by topperc** [comment]
> @topperc thanks for the review, I followed your suggestions.

Yes. Our github configuration will only allow PRs to "squash and merge" so the separate commit won't survive.

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 2 above)
Comment (rzinsly) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:616]: Ok, I created a opened a new one: https://github.com/llvm/llvm-project/pull/117871

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 8]
```diff
@@ -610,22 +610,134 @@ static MCCFIInstruction createDefCFAOffset(const TargetRegisterInfo &TRI,
...
+        .addCFIIndex(CFIIndex)
+        .setMIFlags(MachineInstr::FrameSetup);
+  }
+
+  if (Residual) {
+    RI->adjustReg(MBB, MBBI, DL, SPReg, SPReg, StackOffset::getFixed(-Residual),
+                  MachineInstr::FrameSetup, getStackAlign());
+    if (Residual > ProbeSize) {
+      // s[d|w] zero, 0(sp)
+      bool IsRV64 = STI.hasFeature(RISCV::Feature64Bit);
```
[the remark this answers is not in the document]
Comment (rzinsly) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: Missed that, sorry.
Fixed now.

**Comment by rzinsly** [comment]
Rebased to fix the merge conflicts.

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 9]
```diff
@@ -580,25 +580,134 @@ static MCCFIInstruction createDefCFAOffset(const TargetRegisterInfo &TRI,
...
+            .addCFIIndex(CFIIndex)
+            .setMIFlag(MachineInstr::FrameSetup);
+      }
+    }
+
+    return;
+  }
+
+  // Emit a variable-length allocation probing loop.
+  uint64_t RoundedSize = (Offset / ProbeSize) * ProbeSize;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:656]: is this equivalent to `uint64_t RoundedSize = alignDown(Offset, ProbeSize)`?

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 10]
```diff
@@ -580,25 +580,134 @@ static MCCFIInstruction createDefCFAOffset(const TargetRegisterInfo &TRI,
...
+        MF.addFrameInst(MCCFIInstruction::createDefCfaRegister(nullptr, Reg));
+    BuildMI(MBB, MBBI, DL, TII->get(TargetOpcode::CFI_INSTRUCTION))
+        .addCFIIndex(CFIIndex)
+        .setMIFlags(MachineInstr::FrameSetup);
+  }
+
+  if (Residual) {
+    RI->adjustReg(MBB, MBBI, DL, SPReg, SPReg, StackOffset::getFixed(-Residual),
+                  MachineInstr::FrameSetup, getStackAlign());
+    if (Residual > ProbeSize) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:693]: How can Residual be larger than ProbeSize here?

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 11]
```diff
@@ -1910,3 +2023,72 @@ bool RISCVFrameLowering::isSupportedStackID(TargetStackID::Value ID) const {
...
+  ExitMBB->splice(ExitMBB->end(), &MBB, std::next(MBBI), MBB.end());
+
+  LoopTestMBB->addSuccessor(ExitMBB);
+  LoopTestMBB->addSuccessor(LoopTestMBB);
+  MBB.addSuccessor(LoopTestMBB);
+}
+
+void RISCVFrameLowering::inlineStackProbe(MachineFunction &MF,
+                                          MachineBasicBlock &MBB) const {
+  // Get the instructions that need to be replaced. We emit at most two of
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:2083]: This comment was copied from AArch64 which uses a SmallVector for the "remembering". Is this comment incorrect with respect to the RISC-V code?

File: llvm/lib/Target/RISCV/RISCVMachineFunctionInfo.h [hunk 12]
```diff
@@ -157,6 +159,9 @@ class RISCVMachineFunctionInfo : public MachineFunctionInfo {
 
   bool isVectorCall() const { return IsVectorCall; }
   void setIsVectorCall() { IsVectorCall = true; }
+
+  bool hasStackProbing() const { return StackProbeSize != 0; }
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVMachineFunctionInfo.h:163]: These are unused?

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 10 above)
Comment (rzinsly) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: It can't now, this code will be needed to probe dynamic allocation.
I removed it from this patch.

File: llvm/lib/Target/RISCV/RISCVMachineFunctionInfo.h (hunk 12 above)
Comment (rzinsly) [reply llvm/lib/Target/RISCV/RISCVMachineFunctionInfo.h]: Yes, I removed them now.

File: llvm/test/CodeGen/RISCV/stack-clash-prologue.ll [hunk 13]
```diff
@@ -0,0 +1,537 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -mtriple=riscv64 -mattr=+m -O2 < %s \
+; RUN:   | FileCheck %s -check-prefix=RV64I
+; RUN: llc -mtriple=riscv32 -mattr=+m -O2 < %s \
+; RUN:   | FileCheck %s -check-prefix=RV32I
+
+; Tests copied from PowerPC.
+
+; Free probe
+define i8 @f0() #0 nounwind {
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/stack-clash-prologue.ll:10]: Drop nounwind?

Comment (rzinsly) [reply llvm/test/CodeGen/RISCV/stack-clash-prologue.ll]: Done!

**Review by topperc — APPROVED** [review]
LGTM

File: llvm/test/CodeGen/RISCV/stack-clash-prologue-nounwind.ll (hunk 1 above)
Comment (rzinsly) [reply llvm/test/CodeGen/RISCV/stack-clash-prologue-nounwind.ll]: @lenary hi, is there anything else I should change?

**Review by lenary — APPROVED** [review]
LGTM

**Comment by topperc** [comment]
@rzinsly do you need someone to commit this?

**Comment by rzinsly** [comment]
> @rzinsly do you need someone to commit this?

Yes, please.

**Comment by lenary** [comment]
> > @rzinsly do you need someone to commit this?
> 
> Yes, please.

Please turn off [Keep my email addresses private](https://github.com/settings/emails) setting in your account. Otherwise the commit will be from the github hidden email which we don't use in LLVM. See [LLVM Discourse](https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it) for more information.

**Comment by rzinsly** [comment]
> > > @rzinsly do you need someone to commit this?
> > 
> > 
> > Yes, please.
> 
> Please turn off [Keep my email addresses private](https://github.com/settings/emails) setting in your account. Otherwise the commit will be from the github hidden email which we don't use in LLVM. See [LLVM Discourse](https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it) for more information.

Ok, done!

**Comment by topperc** [comment]
I should have a fix for the buildbot failure in a few minutes

**Comment by rzinsly** [comment]
@topperc I opened a PR to fix it: https://github.com/llvm/llvm-project/pull/119451
But go ahead and push yours.
